[jrl-cmakemodules] Add new port

shixiong2333 opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

🤓 This is a prerequisite for adding the [pinocchio](https://github.com/stack-of-tasks/pinocchio) port (#8928). The next step is to add the [coal](https://github.com/coal-library/coal) port, both of which will utilize [jrl-cmakemodules](https://github.com/jrl-umi3218/jrl-cmakemodules).

🥵 As described in the usage, jrl-cmakemodules need to be imported before defining the project in CMake. However, my testing revealed that some vcpkg-related configurations only take effect after the project is defined. Therefore, I added `project(vcpkg_helper LANGUAGES NONE)` at the beginning of the usage (I will add similar patches to CMakeLists.txt for coal and pinocchio later). I feel this is not a good approach. Is there a better way to write it?

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

<!-- If this PR adds a new port, please uncomment and fill out this checklist: -->

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

<!-- END OF NEW PORT CHECKLIST -->

shixiong2333 committed:
[jrl-cmakemodules] Add new port

shixiong2333 committed:
[supports] !android

shixiong2333 committed:
support sdformat 15

dg0yt commented:
> As described in the usage, jrl-cmakemodules need to be imported before defining the project in CMake. However, my testing revealed that some vcpkg-related configurations only take effect after the project is defined. Therefore, I added `project(vcpkg_helper LANGUAGES NONE)` at the beginning of the usage (I will add similar patches to CMakeLists.txt for coal and pinocchio later). I feel this is not a good approach. Is there a better way to write it?

I suggest you transform your testing to a "test port" `scripts/test_ports/vcpkg-ci-jrl-cmakemodules`. Basically it should try to build a minimal project following closely https://jrl-cmakemodules.readthedocs.io/en/master/pages/base.html#minimal-working-example-1.
It would help understand what doesn't work, and it would help avoiding regressions in the future.

AFAICS jrl-cmake-modules already defers one part of its setup to *after toolchain loading* via https://github.com/jrl-umi3218/jrl-cmakemodules/blob/c16273dd4b38fbf1288669d21f8b6548d9d505a9/base.cmake#L248-L252. Basically this is the key to dealing with steps which need toolchain information.

dg0yt commented on ports/jrl-cmakemodules/vcpkg.json:
```diff
@@ -0,0 +1,18 @@
+  "description": "CMake utility modules used by the stack-of-tasks projects",
+  "homepage": "https://github.com/jrl-umi3218/jrl-cmakemodules",
+  "license": "LGPL-3.0-or-later",
+  "supports": "!android",
```
The android error I see in CI:
~~~
CMake Error at base.cmake:145 (message):
  JRL-CMakemodules require CMake >= 3.22.  Please update your main
  'cmake_minimum_required'
Call Stack (most recent call first):
  CMakeLists.txt:11 (include)
~~~
The project has the right `cmake_minimum_required`. I wonder where this is lost (for the current scope).

shixiong2333 replied:
### Summary: `CMAKE_MINIMUM_REQUIRED_VERSION` changes on Android after `project()`

When building **in an Android CLI environment (NDK toolchain)**,  
`CMAKE_MINIMUM_REQUIRED_VERSION` is **not stable** and can change **after `project()` is called**, even if `cmake_minimum_required()` is set to a higher version at the top level.

#### Minimal reproducer

```cmake
cmake_minimum_required(VERSION 3.22)
message("0 CMAKE_MINIMUM_REQUIRED_VERSION ${CMAKE_MINIMUM_REQUIRED_VERSION}")

set(PROJECT_NAME "jrl-cmakemodules")
set(PROJECT_DESCRIPTION "CMake utility toolbox")

message("1 CMAKE_MINIMUM_REQUIRED_VERSION ${CMAKE_MINIMUM_REQUIRED_VERSION}")
project(${PROJECT_NAME} DESCRIPTION ${PROJECT_DESCRIPTION} LANGUAGES CXX)
message("2 CMAKE_MINIMUM_REQUIRED_VERSION ${CMAKE_MINIMUM_REQUIRED_VERSION}")

if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
  message("3 CMAKE_MINIMUM_REQUIRED_VERSION ${CMAKE_MINIMUM_REQUIRED_VERSION}")
  include(base.cmake)
  message("4 CMAKE_MINIMUM_REQUIRED_VERSION ${CMAKE_MINIMUM_REQUIRED_VERSION}")
endif()
```

#### Output (Android NDK r28c)

```
0 CMAKE_MINIMUM_REQUIRED_VERSION 3.22
1 CMAKE_MINIMUM_REQUIRED_VERSION 3.22
-- The CXX compiler identification is Clang 19.0.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: .../android-ndk-r28c/.../clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
2 CMAKE_MINIMUM_REQUIRED_VERSION 3.10.0
3 CMAKE_MINIMUM_REQUIRED_VERSION 3.10.0
CMake Error at base.cmake:145 (message):
```

#### Key observation

`CMAKE_MINIMUM_REQUIRED_VERSION` changes from **3.22 → 3.10.0** immediately after `project()`.

This behavior:
- ✅ **Only happens on Android**
- ❌ **Does not occur on non-Android toolchains**

---

### Root cause

The Android NDK toolchain explicitly calls:

```cmake
cmake_minimum_required(VERSION 3.10.0)
```

in `android.toolchain.cmake`.

CMake invokes the toolchain **more than once during initial configuration** (notably during compiler detection / try-compile triggered by `project()`).

Relevant excerpt from `android.toolchain.cmake`:

```cmake
cmake_minimum_required(VERSION 3.10.0)

# CMake invokes the toolchain file twice during the first build
# https://github.com/android-ndk/ndk/issues/323
if(ANDROID_NDK_TOOLCHAIN_INCLUDED)
  return()
endif()
set(ANDROID_NDK_TOOLCHAIN_INCLUDED true)
```

Additionally, according to the NDK changelog:

- [Android NDK Issue #2100](https://github.com/android/ndk/issues/2100): Raised minimum required CMake version of the NDK's toolchain
  file from 3.6.0 to 3.10.0 to address warnings from new versions of CMake.

👉 As a result, **the directory variable `CMAKE_MINIMUM_REQUIRED_VERSION` is overwritten by the toolchain**, even if the project set a higher value earlier.

dg0yt commented on ports/jrl-cmakemodules/support_sdformat_15.diff:
```diff
@@ -0,0 +1,11 @@
+ macro(SEARCH_FOR_SDFORMAT)
+   set(
+     SDF_VERSIONS
++    "15"
```
This port doesn't depend on sdformat. Which is probaly fine, because it is a collection of optional helper scripts. But OTOH it means this might be broken by sdformat port updates because the connection is not seen by CI. 
Another thing which might be connected in the test port.
Or modify in a way which would not need maintenance.

shixiong2333 replied:
Additionally, is there a vcpkg command that supports local port testing 🤔 `scripts/test_ports/vcpkg-ci-jrl-cmakemodules-sdformat`

dg0yt commented on ports/jrl-cmakemodules/usage:
```diff
@@ -0,0 +1,25 @@
+    cmake_minimum_required(VERSION 3.22)
+
+    # 1. Bootstrapping: Trigger vcpkg to set paths without detecting compilers yet
+    project(MyProject_vcpkg_helper LANGUAGES NONE)
```
This is not what upstream does.

dg0yt commented on ports/jrl-cmakemodules/usage:
```diff
@@ -0,0 +1,25 @@
+    project(MyProject_vcpkg_helper LANGUAGES NONE)
+
+    # 2. Find the modules
+    find_package(jrl-cmakemodules CONFIG REQUIRED)
```
This is not what upstream does.

And this is what cannot know vcpkg paths before `project()`. It can be solved easily for ports, but needs more thought for user projects.

dg0yt replied:
Thanks. This is something which provides a good hook for dealing with *ports*: The `jrl-cmakemodules_ROOT` CMake variable or environment variable will inform the port where to find the package. The environment variable could be set by a `vcpkg-port-config.cmake` script, so just adding the port dependency would be enough.

For *user projects in manifest mode*, it is not so easy. `project()` doesn't only setup variables, but it is also the step which includes *installing the dependencies*. OTOH user projects could just use plain FetchContent.

shixiong2333 commented:
> > As described in the usage, jrl-cmakemodules need to be imported before defining the project in CMake. However, my testing revealed that some vcpkg-related configurations only take effect after the project is defined. Therefore, I added `project(vcpkg_helper LANGUAGES NONE)` at the beginning of the usage (I will add similar patches to CMakeLists.txt for coal and pinocchio later). I feel this is not a good approach. Is there a better way to write it?
> 
> I suggest you transform your testing to a "test port" `scripts/test_ports/vcpkg-ci-jrl-cmakemodules`. Basically it should try to build a minimal project following closely https://jrl-cmakemodules.readthedocs.io/en/master/pages/base.html#minimal-working-example-1. It would help understand what doesn't work, and it would help avoiding regressions in the future.
> 
> AFAICS jrl-cmake-modules already defers one part of its setup to _after toolchain loading_ via https://github.com/jrl-umi3218/jrl-cmakemodules/blob/c16273dd4b38fbf1288669d21f8b6548d9d505a9/base.cmake#L248-L252. Basically this is the key to dealing with steps which need toolchain information.

- added  `scripts/test_ports/vcpkg-ci-jrl-cmakemodules`
- I tried moving `project(...)` before `find_package(jrl-cmakemodules CONFIG REQUIRED)`, and then explicitly `include("${CMAKE_CURRENT_LIST_DIR}/post-project.cmake")`, but this resulted in some other inconsistencies in pinocchio, such as some project-related variables in jrl not being set correctly.
- The usage I used is, in my opinion, the least problematic and tested approach for subsequent pinocchio/coal changes.
- Unless there's a way to find the path to jrl-cmakemodules before `project(....)` in a vcpkg environment.

Based on my current knowledge, these are the only suggestions I can offer ~

shixiong2333 committed:
Add test_ports for jrl-cmakemodules

dg0yt commented:
> Unless there's a way to find the path to jrl-cmakemodules before `project(....)` in a vcpkg environment.

As I said before, this can be handled easily for ports. Dependencies are installed before configuration, and we control input variables and environment.

I believe using `project()` twice in the same scope should not be encouraged. This engages the combination of the quirks of vcpkg and jrl-cmakemodules when it comes to side effects of `project()`. Instead the `jrl-cmakemodules` should be asked to offer a mode which can be used safely after `project()`.

shixiong2333 committed:
Add vcpkg-port-config.cmake for jrl-cmakemodules

shixiong2333 committed:
change vcpkg-ci-jrl-cmakemodules-sdformat

dg0yt commented on scripts/test_ports/vcpkg-ci-jrl-cmakemodules-sdformat/portfile.cmake:
```diff
@@ -0,0 +1,9 @@
+vcpkg_cmake_configure(
+    SOURCE_PATH "${CURRENT_PORT_DIR}/project"
+    OPTIONS
+        -Djrl-cmakemodules_DIR=${jrl-cmakemodules_DIR}
```
```suggestion
```

dg0yt commented on scripts/test_ports/vcpkg-ci-jrl-cmakemodules-sdformat/project/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+include(${jrl-cmakemodules_DIR}/base.cmake)
+
+# 4. Setup project arguments (Version, Policies, etc.)
+COMPUTE_PROJECT_ARGS(PROJECT_ARGS LANGUAGES CXX)
```
For readability:
```suggestion
compute_project_args(PROJECT_ARGS LANGUAGES CXX)
```

dg0yt commented on scripts/test_ports/vcpkg-ci-jrl-cmakemodules-sdformat/project/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+project(${PROJECT_NAME} ${PROJECT_ARGS})
+
+# 5. Find sdformat
+include(${jrl-cmakemodules_DIR}/sdformat.cmake)
```
Isn't this implicitly included?

dg0yt reviewed (commented):
(There more I see from it, the more I would recommend to avoid it. People should learn and use Modern CMake, not another non-standard language on top of it...)

First I didn't notice that you added two test ports. Please apply my suggestions to the main test port. In the end it is probably enough to have one test port with an `sdformat` default-feature. This would still allow for cheap local testing with vcpkg-ci-jrl-cmakemodules[core]. But simply keep two ports if it is easier for you now.

shixiong2333 commented:
> (There more I see from it, the more I would recommend to avoid it. People should learn and use Modern CMake, not another non-standard language on top of it...)

Completely agree ! Just to ensure pinocchio can be properly imported later. Don't recommend users use jrl-cmakemodules.

shixiong2333 commented on scripts/test_ports/vcpkg-ci-jrl-cmakemodules-sdformat/project/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+find_package(jrl-cmakemodules CONFIG REQUIRED)
+
+# 3. Include the base infrastructure
+include(${jrl-cmakemodules_DIR}/base.cmake)
```
Test results show that this is true; otherwise, there would be problems using the cmake functions related to jrl-cmakemodules.

shixiong2333 committed:
set env jrl-cmakemodules_ROOT

shixiong2333 committed:
change vcpkg-port-config.cmake

shixiong2333 commented on ports/jrl-cmakemodules/vcpkg-port-config.cmake:
```diff
@@ -0,0 +1,2 @@
+include_guard(GLOBAL)
+set(jrl-cmakemodules_DIR "${CMAKE_CURRENT_LIST_DIR}")
```
Perfect, it works!

shixiong2333 committed:
Add JRL_SKIP_FIND_PYTHON_COMPONENTS
